### PR TITLE
Improve typing in expressions module.

### PR DESCRIPTION
**Pull request by jpinner-lyft**

### CODE REVIEW

File: pynamodb/expressions/projection.py [hunk 1]
```diff
@@ -1,16 +1,20 @@
...
+from typing import List
+from typing import Union
+
 from pynamodb.attributes import Attribute
 from pynamodb.expressions.operand import Path
 from pynamodb.expressions.util import substitute_names
 
 
-def create_projection_expression(attributes_to_get, placeholders):
+def create_projection_expression(attributes_to_get, placeholders: Dict[str, str]) -> str:
```
Comment (ikonst) [inline pynamodb/expressions/projection.py:10]: `attributes_to_get` too?

File: pynamodb/expressions/update.py [hunk 2]
```diff
@@ -86,22 +96,24 @@ def add_action(self, action: Action) -> None:
...
-        expression = self._add_clause(expression, 'DELETE', self.delete_actions, placeholder_names, expression_attribute_values)
-        return expression
+    def serialize(self, placeholder_names: Dict[str, str], expression_attribute_values: Dict[str, str]) -> Optional[str]:
+        clauses = [
+            self._get_clause('SET', self.set_actions, placeholder_names, expression_attribute_values),
+            self._get_clause('REMOVE', self.remove_actions, placeholder_names, expression_attribute_values),
+            self._get_clause('ADD', self.add_actions, placeholder_names, expression_attribute_values),
+            self._get_clause('DELETE', self.delete_actions, placeholder_names, expression_attribute_values),
+        ]
+        expression = ' '.join([clause for clause in clauses if clause is not None])
```
Comment (ikonst) [inline pynamodb/expressions/update.py:106]:
```suggestion
        expression = ' '.join(clause for clause in clauses if clause is not None)
```

File: pynamodb/expressions/update.py [hunk 3]
```diff
@@ -86,22 +96,24 @@ def add_action(self, action: Action) -> None:
...
-        actions = ", ".join([action.serialize(placeholder_names, expression_attribute_values) for action in actions])
-        return keyword + " " + actions if actions else None
+    def _get_clause(
+            keyword: str,
+            actions: Sequence[Action],
+            placeholder_names: Dict[str, str],
+            expression_attribute_values: Dict[str, str]
+    ) -> Optional[str]:
+        actions_string = ', '.join(
+            [action.serialize(placeholder_names, expression_attribute_values) for action in actions]
```
Comment (ikonst) [inline pynamodb/expressions/update.py:117]:
```suggestion
            action.serialize(placeholder_names, expression_attribute_values) for action in actions
```

File: pynamodb/expressions/condition.py [hunk 4]
```diff
@@ -44,10 +44,6 @@ def __repr__(self):
         values = [str(value) for value in self.values]
         return self.format_string.format(*values, operator=self.operator)
 
-    def __nonzero__(self):
```
Comment (ikonst) [inline pynamodb/expressions/condition.py:47]: good catch

File: pynamodb/expressions/update.py (hunk 2 above)
Comment (jpinner-lyft) [reply pynamodb/expressions/update.py]: :+1:

File: pynamodb/expressions/update.py (hunk 3 above)
Comment (jpinner-lyft) [reply pynamodb/expressions/update.py]: :+1:

File: pynamodb/expressions/projection.py (hunk 1 above)
Comment (jpinner-lyft) [reply pynamodb/expressions/projection.py:10]: that comes from connection base where it's typed as `Optional[Any]` -- so far I've inferred its:
`Union[List[Attribute], List[Path], List[str], Attribute, Path, str]`
but I was going to punt on it for now

**Review by ikonst — APPROVED** [review]

**Pull request merged**